### ISSUE TITLE
Grace notes length

### ISSUE THREAD

**Issue by musikboden** [paulrosen/abcjs#458]
I've noticed that all grace notes sounding lengths are dependant on the note after the grace notes, is it supposed to be like that or is it a bug?

In this case the 2 grace notes will have the length of a quarter note each, so they will occupy half of the whole note length.
![Skärmavbild 2020-06-24 kl  09 54 37](https://user-images.githubusercontent.com/18282700/85519636-16fb9e00-b602-11ea-9825-4d451323010e.png)

Also here the grace note will have a length of a half note

![Skärmavbild 2020-06-24 kl  09 55 27](https://user-images.githubusercontent.com/18282700/85520669-87ef8580-b603-11ea-8cdb-40fd927da890.png)

and here they will sound as 3 quarter triplets
![Skärmavbild 2020-06-24 kl  10 20 19](https://user-images.githubusercontent.com/18282700/85521506-9f7b3e00-b604-11ea-8447-58de38d720ee.png)

Setting visual note length to grace notes is sometimes working and sometimes not.
A single grace note will not be affected
![Skärmavbild 2020-06-24 kl  09 58 13](https://user-images.githubusercontent.com/18282700/85522184-98a0fb00-b605-11ea-9515-48a90ba57c1a.png)

but several grace notes will be affected to a certain degree, doubling and halving will work fine
![Skärmavbild 2020-06-24 kl  10 35 01](https://user-images.githubusercontent.com/18282700/85522788-6643cd80-b606-11ea-82b0-e25f973fd8cc.png)

but making a dotted will not work
![Skärmavbild 2020-06-24 kl  10 00 19](https://user-images.githubusercontent.com/18282700/85523020-b6bb2b00-b606-11ea-9115-c221874c66b3.png)

It is the same effect on both acciaccaturas as appoggiaturas

http://abcnotation.com/wiki/abc:standard:v2.1#grace_notes

**Comment by paulrosen** [comment]
Yeah - the current algorithm doesn't work. What's your suggestion for how to figure out the length of them? 

It needs to work at both slow and fast tempos and with note lengths of sixteenth to whole notes. And it needs to work whether this is one grace note or many. And it might need to be some kind of compromise for different styles of music.

**Comment by musikboden** [comment]
This seems to be very complex.
This is taken from "abc2midi" 
https://ifdo.ca/~seymour/runabc/abcguide/abc2midi_guide.html

> Grace Notes

>%%MIDI grace a/b
>sets the fraction of the next note that grace notes will take up. a must be between 1 and b-1. The >grace notes may not sound natural in this approach, since the length of the individual grace notes >varies with the complexity of the grace and the length of the following note. Here is a sample abc >file.


>X:1
>T: grace fraction
>M: 2/4
>L: 1/8
>K: C
>%%MIDI grace 1/4
>{E}D2 {GFE}D2| {DCDC}C4|
>%%MIDI grace 1/2
>{E}D2 {GFE}D2| {DCDC}C4|
>%%MIDI grace 3/4
>{E}D2 {GFE}D2| {DCDC}C4|
>which sounds like this.
>A different approach (which is now the default) assumes that the grace notes always have fixed >duration specified by a fraction of the unit length. To use the other approach you would specify

>%%MIDI gracedivider b
>where b specifies how many parts to divide the unit length specified by the L: field command. For >example, if b = 4 and L: = 1/8, then every grace note would be 1/(8*4) or a 32nd note. Time would >be stolen from the note to which the grace note is applied. If that note is not long enough to handle >the grace sequence, then the grace notes would be assigned 0 duration. Here is an example.

>X:1
>T: grace divider
>M: 2/4
>L: 1/8
>K: C
>%%MIDI gracedivider 4
>{E}D2 {GFE}D2| {DCDC}C4|
>%%MIDI gracedivider 2
>{E}D2 {GFE}D2| {DCDC}C4|

>This sounds a little better.
>Note that the MIDI command has no effect on grace notes introduced using the M, R, T, !trill!, >!mordent! indications.

>In some cases, you may want abc2midi to ignore all grace notes markings. To do this, add the ->NGRA (no grace notes) to the runtime parameter list.

I made a test with "Dorico" notation software and it will handle the grace notes differently.
The length of the grace note will be taken from note BEFORE the GN, so that the note efter GN will be untouched, and the note before the GN will be shortened.
I find this way to be the most "musical" way to do it (I should say that I'm not familiar with all kind of music styles though and maybe it would be good if there was a command to switch between the 2 ways of doing it?)

this was my test mtrl:
![Skärmavbild 2020-06-27 kl  22 54 25](https://user-images.githubusercontent.com/18282700/85942942-11b59080-b92d-11ea-8463-0cbed0f8e391.png)

When using quarter = 60 bpm
all note examples resulted in grace note length = 80/240 of a sixteenth note /each

When using quarter = 120 bpm
all note examples except the last example, resulted in grace note length = 160/240 of a sixteenth note/each
the last example with 9 grace notes resulted in grace note length = 106/240 of a sixteenth/each

When using quarter = 240 bpm
the first 10 examples (bar 21 - 28) resulted in grace note length = 320/240 of a sixteenth/each
bar 29 first example (4GN) grace note length = 240/240 of a sixteenth note/each
second example (5GN), grace note length = 192/240....
bar 30 (9GN), grace note length = 106/240...etc

So the grace note length is depended on the tempo and, the note length of the note BEFORE the grace note/notes

So if you want to follow the Dorico example, there is a formula for this, I will explore what happens with the grace notes length, and also with the note before the grace notes, I think it will be shortened by the amount of the grace notes lengths plus a gap (rest).

I will think some more on this, and it would be great if more people could chime in and give some examples of how different music styles handle grace notes.
Before beat, at the beat, length etc.

**Comment by paulrosen** [comment]
Thanks for that analysis! When I looked at it before it just seemed very idiosyncratic. 

My first inclination is that tempo shouldn't be a factor - that is, the music should be the same at any tempo, but I think that isn't realistic.

I don't have much to add about how it should be, except that I understand that Scottish bagpipes start their gracenotes on the beat instead of before the beat.

There may need to be lots of user parameters for this. Perhaps the first pass should be for every variable to be specified as an input parameter. Then it would be easier to experiment. I'm not sure I could even list all the factors without studying it some more.

**Comment by musikboden** [comment]
My thoughts on this is:
*The grace note/s should have the same length irrespective of tempo, as long as the next note, or the note before, is long enough to make room for the grace note sequence.
If the next following note, or note before the sequence, is note long enough, the grace notes should be shortened so much they would fit the length.

*I think you're right about implementing some input parameters, and that the first would be the decision to chose wether the grace notes should come on the beat or before, what do you think?
